agent fails when run outside a git repository

When the agent is run outside of a git repository, it fails when trying to collect metadata:
```
Exception in thread "main" java.lang.reflect.InvocationTargetException
      at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
      at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
      at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
      at java.base/java.lang.reflect.Method.invoke(Method.java:566)
      at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:513)
      at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:525)
Caused by: java.lang.IllegalArgumentException: One of setGitDir or setWorkTree must be called.
      at com.appland.shade.org.eclipse.jgit.lib.BaseRepositoryBuilder.requireGitDirOrWorkTree(BaseRepositoryBuilder.java:636)
      at com.appland.shade.org.eclipse.jgit.lib.BaseRepositoryBuilder.setup(BaseRepositoryBuilder.java:600)
      at com.appland.shade.org.eclipse.jgit.storage.file.FileRepositoryBuilder.build(FileRepositoryBuilder.java:55)
      at com.appland.appmap.util.GitUtil.open(GitUtil.java:65)
      at com.appland.appmap.util.GitUtil.open(GitUtil.java:53)
      at com.appland.appmap.util.GitUtil.findSourceRoots(GitUtil.java:158)
      at com.appland.appmap.Agent.premain(Agent.java:63)
      ... 6 more
FATAL ERROR in native method: processing of -javaagent failed
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -40,7 +40,7 @@ jobs:
       source "$HOME/.asdf/asdf.sh"
       asdf install java latest:adoptopenjdk-8
       asdf prune java adoptopenjdk-8
-      asdf local java latest:adoptopenjdk-8
+      asdf global java latest:adoptopenjdk-8
 
   - name: jdk 11
     env:
@@ -50,7 +50,7 @@ jobs:
       source "$HOME/.asdf/asdf.sh"
       asdf install java latest:adoptopenjdk-11
       asdf prune java adoptopenjdk-11
-      asdf local java latest:adoptopenjdk-11
+      asdf global java latest:adoptopenjdk-11
   - name: jdk 17
     env:
     - JDK=17
@@ -59,7 +59,7 @@ jobs:
       source "$HOME/.asdf/asdf.sh"
       asdf install java latest:adoptopenjdk-17
       asdf prune java adoptopenjdk-17
-      asdf local java latest:adoptopenjdk-17
+      asdf global java latest:adoptopenjdk-17
 
 script:
 - |

--- a/agent/build.gradle
+++ b/agent/build.gradle
@@ -9,7 +9,6 @@
 import com.appland.tasks.ShadowRelocation
 
 plugins {
-  id 'java-library'
   id 'jacoco'
   id 'signing'
 }

--- a/agent/config/checkstyle/main/java/include.txt
+++ b/agent/config/checkstyle/main/java/include.txt
@@ -1,0 +1,6 @@
+com/appland/appmap/config/AppMapConfig.java
+com/appland/appmap/Agent.java
+com/appland/appmap/output/v1/CodeObject.java
+com/appland/appmap/record/AppMapSerializer.java
+com/appland/appmap/record/RecordingSession.java
+com/appland/appmap/util/GitUtil.java

--- a/agent/src/main/java/com/appland/appmap/output/v1/CodeObject.java
+++ b/agent/src/main/java/com/appland/appmap/output/v1/CodeObject.java
@@ -66,8 +66,9 @@ public class CodeObject {
   @JSONField(name = "location")
   public String getLocation() {
     String ret = this.file != null? this.file : null;
-    if (ret != null && this.lineno != null)
+    if (ret != null && this.lineno != null) {
       ret += ":" + this.lineno;
+    }
 
     return ret;
   }
@@ -425,12 +426,14 @@ public class CodeObject {
 
   private boolean equalBySubstring(String s1, String s2, int start, int end) {
     int sublen = end - start;
-    if (s1.length() != sublen)
+    if (s1.length() != sublen) {
       return false;
+    }
 
     for (int idx = 0; idx < sublen; idx++) {
-      if (s1.charAt(idx) != s2.charAt(start + idx))
+      if (s1.charAt(idx) != s2.charAt(start + idx)) {
         return false;
+      }
     }
 
     return true;

--- a/agent/src/main/java/com/appland/appmap/record/RecordingSession.java
+++ b/agent/src/main/java/com/appland/appmap/record/RecordingSession.java
@@ -146,8 +146,9 @@ public class RecordingSession {
       String[] parts = key.split(":");
 
       CodeObject methodBranch = registeredObjects.getMethodBranch(parts[0], parts[1], Boolean.valueOf(parts[2]), Integer.valueOf(parts[3]));
-      if (methodBranch != null)
+      if (methodBranch != null) {
         classMap.add(methodBranch);
+      }
     }
 
     return classMap;

--- a/agent/src/main/java/com/appland/appmap/util/GitUtil.java
+++ b/agent/src/main/java/com/appland/appmap/util/GitUtil.java
@@ -21,6 +21,7 @@ import org.eclipse.jgit.api.errors.GitAPIException;
 import org.eclipse.jgit.errors.NoWorkTreeException;
 import org.eclipse.jgit.errors.RevisionSyntaxException;
 import org.eclipse.jgit.lib.Constants;
+import org.eclipse.jgit.lib.ObjectId;
 import org.eclipse.jgit.lib.Repository;
 import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
 import org.eclipse.jgit.transport.RemoteConfig;
@@ -35,6 +36,8 @@ public class GitUtil implements AutoCloseable {
   private static final TaggedLogger logger = AppMapConfig.getLogger(null);
 
   private Git git;
+  private Path fsBase;
+  private ObjectId tree; // HEAD^{tree}
 
   // Map from the Git path of a source root (that starts with
   // src/{main,test}/java) to its real location in the filesystem.
@@ -45,23 +48,23 @@ public class GitUtil implements AutoCloseable {
   // failed.
   private static Map<String, Optional<String>> sourcePaths = new HashMap<>();
 
-  private GitUtil(Git git) {
+  private GitUtil(Git git, ObjectId tree, Path fsBase) {
     this.git = git;
+    this.tree = tree;
+    this.fsBase = fsBase;
   }
 
   public static GitUtil open() throws IOException {
-    return open(null);
-  }
-
-  public static GitUtil open(File workTree) throws IOException {
     try {
       FileRepositoryBuilder builder = new FileRepositoryBuilder()
           .readEnvironment();
-      if (workTree != null) {
-        builder.setWorkTree(workTree);
-      } else {
+
         builder.findGitDir();
-      }
+        if (builder.getGitDir() == null) {
+          logger.debug("Working directory {}, not in a git repo", () -> Paths.get("").toAbsolutePath());
+          return null;
+        }
+
       Repository repository = builder.build();
       if (repository.isBare()) {
         logger.warn(
@@ -69,7 +72,14 @@ public class GitUtil implements AutoCloseable {
         return null;
       }
 
-      return new GitUtil(new Git(repository));
+      Path fsBase = AppMapConfig.get().configFile.toAbsolutePath().getParent();
+      ObjectId tree = repository.resolve(Constants.HEAD + "^{tree}");
+      if (tree == null) {
+        logger.warn("Couldn't resolve HEAD to a tree in {}, source paths may be incorrect", fsBase);
+        return null;
+      }
+
+      return new GitUtil(new Git(repository), tree, fsBase);
     } catch (IOException e) {
       logger.warn(e);
     }
@@ -142,6 +152,7 @@ public class GitUtil implements AutoCloseable {
   private static final byte[] GIT_TEST_JAVA = "src/test/java".getBytes();
   private static int GIT_TEST_JAVA_LEN = GIT_TEST_JAVA.length;
 
+
   /**
    * If we're in a git repo, walk all the directories under the directory that
    * contains the config, looking for source directories (i.e. those that start
@@ -152,8 +163,6 @@ public class GitUtil implements AutoCloseable {
    * @throws IOException if something goes wrong accessing the git repo
    */
   public static void findSourceRoots() throws IOException {
-    // Start the search for source roots in the directory containing the config.
-    Path fsBase = AppMapConfig.get().configFile.toAbsolutePath().getParent();
 
     try (GitUtil git = GitUtil.open()) {
       if (git == null) {
@@ -162,10 +171,10 @@ public class GitUtil implements AutoCloseable {
 
       Repository repository = git.getRepository();
       Path fsRoot = repository.getWorkTree().toPath();
-      String gitBase = fsRoot.relativize(fsBase).toString().replace(File.separator, "/");
+      String gitBase = fsRoot.relativize(git.fsBase).toString().replace(File.separator, "/");
       logger.debug("repoRoot: {}, gitCwd: {}", fsRoot, gitBase);
       try (TreeWalk treeWalk = new TreeWalk(repository)) {
-        treeWalk.addTree(repository.resolve(Constants.HEAD + "^{tree}"));
+        treeWalk.addTree(git.tree);
 
         if (gitBase.length() > 0) {
           // Ignore the parts of the repo that aren't under the base directory.

--- a/agent/test/access/access.bats
+++ b/agent/test/access/access.bats
@@ -6,7 +6,8 @@ load '../helper'
 
 sep="$JAVA_PATH_SEPARATOR"
 appmap_jar="$(find_agent_jar)"
-test_cp="test/access${sep}build/classes/java/test"
+wd="$(git rev-parse --show-toplevel)"/agent
+test_cp="${wd}/test/access${sep}${wd}/build/classes/java/test"
 java_cmd="java -javaagent:'${appmap_jar}' -cp '${test_cp}'"
 
 setup() {
@@ -29,3 +30,23 @@ setup() {
   eval "$cmd" | jq -e '.events[3] | select(.event=="call" and .method_id=="myPrivateMethod")'
 }
 
+@test "outside git repo" {
+  cp appmap.yml "$BATS_TEST_TMPDIR"/.
+  cd "$BATS_TEST_TMPDIR"
+
+  # sanity check
+  run git rev-parse --show-top-level
+  assert_output -p 'not a git repository'
+  assert_failure
+
+  local cmd="${java_cmd} RecordPackage"
+  run bash -c "eval \"$cmd\" 2>/dev/null"
+  assert_success
+  local recording="${output}"
+  run jq -e '.events | length | select(. == 4)' <<< "${recording}"
+  assert_success
+
+  # metadata shouldn't have any git info
+  run jq '.metadata.git' <<< "${recording}"
+  assert_output 'null'
+}

--- a/annotation/build.gradle
+++ b/annotation/build.gradle
@@ -7,7 +7,6 @@
  */
 
 plugins {
-  id 'java-library'
   id 'jacoco'
   id 'signing'
 }

--- a/build.gradle
+++ b/build.gradle
@@ -40,3 +40,29 @@ nexusPublishing {
         }
     }
 }
+
+subprojects {
+  apply plugin: 'java-library'
+  apply plugin: 'checkstyle'
+
+  checkstyle {
+    toolVersion = "9.3"
+    configFile = rootProject.file("config/checkstyle/checkstyle.xml")
+  }
+
+  tasks.named('checkstyleMain').configure {
+      def included = project.file("config/checkstyle/main/java/include.txt")
+      if (included.exists()) {
+        logger.debug("checkstyle for ${project.name}, include file: ${included}")
+        def includedFiles = included.readLines();
+        logger.debug("checking files ${includedFiles}")
+        include includedFiles
+        inputs.files(includedFiles)
+      }
+
+      reports {
+          html.enabled = false // Disable HTML report
+          xml.enabled = true  // Disable XML report if you don't want it
+      }
+  }
+}

--- a/buildSrc/build.gradle
+++ b/buildSrc/build.gradle
@@ -6,7 +6,7 @@ buildscript {
     }
   }
   dependencies {
-    classpath 'com.github.jengelman.gradle.plugins:shadow:5.2.0'
+    classpath 'gradle.plugin.com.github.johnrengelman:shadow:7.1.2'
   }
 }
 
@@ -23,7 +23,7 @@ repositories {
   }
 
 dependencies {
-  implementation 'com.github.jengelman.gradle.plugins:shadow:5.2.0'
+  implementation 'gradle.plugin.com.github.johnrengelman:shadow:7.1.2'
 }
 
 

--- a/config/checkstyle/checkstyle.xml
+++ b/config/checkstyle/checkstyle.xml
@@ -1,0 +1,366 @@
+<?xml version="1.0"?>
+<!DOCTYPE module PUBLIC "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN" "https://checkstyle.org/dtds/configuration_1_3.dtd">
+
+<!--
+    Checkstyle configuration that checks the Google coding conventions from Google Java Style
+    that can be found at https://google.github.io/styleguide/javaguide.html
+
+    Checkstyle is very configurable. Be sure to read the documentation at
+    http://checkstyle.org (or in your downloaded distribution).
+
+    To completely disable a check, just comment it out or delete it from the file.
+    To suppress certain violations please review suppression filters.
+
+    Authors: Max Vetrenko, Ruslan Diachenko, Roman Ivanov.
+ -->
+
+<module name="Checker">
+  <property name="charset" value="UTF-8" />
+
+  <property name="severity" value="ignore" />
+
+  <property name="fileExtensions" value="java, properties, xml" />
+  <!-- Excludes all 'module-info.java' files              -->
+  <!-- See https://checkstyle.org/config_filefilters.html -->
+  <module name="BeforeExecutionExclusionFileFilter">
+    <property name="fileNamePattern" value="module\-info\.java$" />
+  </module>
+  <!-- https://checkstyle.org/config_filters.html#SuppressionFilter -->
+  <module name="SuppressionFilter">
+    <property name="file" value="${org.checkstyle.google.suppressionfilter.config}"
+      default="checkstyle-suppressions.xml" />
+    <property name="optional" value="true" />
+  </module>
+
+  <!-- Checks for whitespace                               -->
+  <!-- See http://checkstyle.org/config_whitespace.html -->
+  <module name="FileTabCharacter">
+    <property name="eachLine" value="true" />
+  </module>
+
+  <module name="LineLength">
+    <property name="fileExtensions" value="java" />
+    <property name="max" value="100" />
+    <property name="ignorePattern" value="^package.*|^import.*|a href|href|http://|https://|ftp://" />
+  </module>
+
+  <module name="TreeWalker">
+    <module name="OuterTypeFilename" />
+    <module name="IllegalTokenText">
+      <property name="tokens" value="STRING_LITERAL, CHAR_LITERAL" />
+      <property name="format"
+        value="\\u00(09|0(a|A)|0(c|C)|0(d|D)|22|27|5(C|c))|\\(0(10|11|12|14|15|42|47)|134)" />
+      <property name="message"
+        value="Consider using special escape sequence instead of octal value or Unicode escaped value." />
+    </module>
+    <module name="AvoidEscapedUnicodeCharacters">
+      <property name="allowEscapesForControlCharacters" value="true" />
+      <property name="allowByTailComment" value="true" />
+      <property name="allowNonPrintableEscapes" value="true" />
+    </module>
+    <module name="AvoidStarImport" />
+    <module name="OneTopLevelClass" />
+    <module name="NoLineWrap">
+      <property name="tokens" value="PACKAGE_DEF, IMPORT, STATIC_IMPORT" />
+    </module>
+    <module name="EmptyBlock">
+      <property name="option" value="TEXT" />
+      <property name="tokens"
+        value="LITERAL_TRY, LITERAL_FINALLY, LITERAL_IF, LITERAL_ELSE, LITERAL_SWITCH" />
+    </module>
+    <module name="NeedBraces">
+      <property name="severity" value="error" />
+      <property name="tokens"
+        value="LITERAL_DO, LITERAL_ELSE, LITERAL_FOR, LITERAL_IF, LITERAL_WHILE" />
+    </module>
+    <module name="LeftCurly">
+      <property name="tokens"
+        value="ANNOTATION_DEF, CLASS_DEF, CTOR_DEF, ENUM_CONSTANT_DEF, ENUM_DEF,
+                    INTERFACE_DEF, LAMBDA, LITERAL_CASE, LITERAL_CATCH, LITERAL_DEFAULT,
+                    LITERAL_DO, LITERAL_ELSE, LITERAL_FINALLY, LITERAL_FOR, LITERAL_IF,
+                    LITERAL_SWITCH, LITERAL_SYNCHRONIZED, LITERAL_TRY, LITERAL_WHILE, METHOD_DEF,
+                    OBJBLOCK, STATIC_INIT, RECORD_DEF, COMPACT_CTOR_DEF" />
+    </module>
+    <module name="RightCurly">
+      <property name="id" value="RightCurlySame" />
+      <property name="tokens"
+        value="LITERAL_TRY, LITERAL_CATCH, LITERAL_FINALLY, LITERAL_IF, LITERAL_ELSE,
+                    LITERAL_DO" />
+    </module>
+    <module name="RightCurly">
+      <property name="id" value="RightCurlyAlone" />
+      <property name="option" value="alone" />
+      <property name="tokens"
+        value="CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE, STATIC_INIT,
+                    INSTANCE_INIT, ANNOTATION_DEF, ENUM_DEF, INTERFACE_DEF, RECORD_DEF,
+                    COMPACT_CTOR_DEF" />
+    </module>
+    <module name="SuppressionXpathSingleFilter">
+      <!-- suppresion is required till https://github.com/checkstyle/checkstyle/issues/7541 -->
+      <property name="id" value="RightCurlyAlone" />
+      <property name="query"
+        value="//RCURLY[parent::SLIST[count(./*)=1]
+                                     or preceding-sibling::*[last()][self::LCURLY]]" />
+    </module>
+    <module name="WhitespaceAfter">
+      <property name="tokens"
+        value="COMMA, SEMI, TYPECAST, LITERAL_IF, LITERAL_ELSE,
+                    LITERAL_WHILE, LITERAL_DO, LITERAL_FOR, DO_WHILE" />
+    </module>
+    <module name="WhitespaceAround">
+      <property name="allowEmptyConstructors" value="true" />
+      <property name="allowEmptyLambdas" value="true" />
+      <property name="allowEmptyMethods" value="true" />
+      <property name="allowEmptyTypes" value="true" />
+      <property name="allowEmptyLoops" value="true" />
+      <property name="ignoreEnhancedForColon" value="false" />
+      <property name="tokens"
+        value="ASSIGN, BAND, BAND_ASSIGN, BOR, BOR_ASSIGN, BSR, BSR_ASSIGN, BXOR,
+                    BXOR_ASSIGN, COLON, DIV, DIV_ASSIGN, DO_WHILE, EQUAL, GE, GT, LAMBDA, LAND,
+                    LCURLY, LE, LITERAL_CATCH, LITERAL_DO, LITERAL_ELSE, LITERAL_FINALLY,
+                    LITERAL_FOR, LITERAL_IF, LITERAL_RETURN, LITERAL_SWITCH, LITERAL_SYNCHRONIZED,
+                    LITERAL_TRY, LITERAL_WHILE, LOR, LT, MINUS, MINUS_ASSIGN, MOD, MOD_ASSIGN,
+                    NOT_EQUAL, PLUS, PLUS_ASSIGN, QUESTION, RCURLY, SL, SLIST, SL_ASSIGN, SR,
+                    SR_ASSIGN, STAR, STAR_ASSIGN, LITERAL_ASSERT, TYPE_EXTENSION_AND" />
+      <message key="ws.notFollowed"
+        value="WhitespaceAround: ''{0}'' is not followed by whitespace. Empty blocks may only be represented as '{}' when not part of a multi-block statement (4.1.3)" />
+      <message key="ws.notPreceded"
+        value="WhitespaceAround: ''{0}'' is not preceded with whitespace." />
+    </module>
+    <module name="OneStatementPerLine" />
+    <module name="MultipleVariableDeclarations" />
+    <module name="ArrayTypeStyle" />
+    <module name="MissingSwitchDefault" />
+    <module name="FallThrough" />
+    <module name="UpperEll" />
+    <module name="ModifierOrder" />
+    <module name="EmptyLineSeparator">
+      <property name="tokens"
+        value="PACKAGE_DEF, IMPORT, STATIC_IMPORT, CLASS_DEF, INTERFACE_DEF, ENUM_DEF,
+                    STATIC_INIT, INSTANCE_INIT, METHOD_DEF, CTOR_DEF, VARIABLE_DEF, RECORD_DEF,
+                    COMPACT_CTOR_DEF" />
+      <property name="allowNoEmptyLineBetweenFields" value="true" />
+    </module>
+    <module name="SeparatorWrap">
+      <property name="id" value="SeparatorWrapDot" />
+      <property name="tokens" value="DOT" />
+      <property name="option" value="nl" />
+    </module>
+    <module name="SeparatorWrap">
+      <property name="id" value="SeparatorWrapComma" />
+      <property name="tokens" value="COMMA" />
+      <property name="option" value="EOL" />
+    </module>
+    <module name="SeparatorWrap">
+      <!-- ELLIPSIS is EOL until https://github.com/google/styleguide/issues/259 -->
+      <property name="id" value="SeparatorWrapEllipsis" />
+      <property name="tokens" value="ELLIPSIS" />
+      <property name="option" value="EOL" />
+    </module>
+    <module name="SeparatorWrap">
+      <!-- ARRAY_DECLARATOR is EOL until https://github.com/google/styleguide/issues/258 -->
+      <property name="id" value="SeparatorWrapArrayDeclarator" />
+      <property name="tokens" value="ARRAY_DECLARATOR" />
+      <property name="option" value="EOL" />
+    </module>
+    <module name="SeparatorWrap">
+      <property name="id" value="SeparatorWrapMethodRef" />
+      <property name="tokens" value="METHOD_REF" />
+      <property name="option" value="nl" />
+    </module>
+    <module name="PackageName">
+      <property name="format" value="^[a-z]+(\.[a-z][a-z0-9]*)*$" />
+      <message key="name.invalidPattern"
+        value="Package name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="TypeName">
+      <property name="tokens"
+        value="CLASS_DEF, INTERFACE_DEF, ENUM_DEF,
+                    ANNOTATION_DEF, RECORD_DEF" />
+      <message key="name.invalidPattern"
+        value="Type name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="MemberName">
+      <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9]*$" />
+      <message key="name.invalidPattern"
+        value="Member name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="ParameterName">
+      <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$" />
+      <message key="name.invalidPattern"
+        value="Parameter name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="LambdaParameterName">
+      <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$" />
+      <message key="name.invalidPattern"
+        value="Lambda parameter name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="CatchParameterName">
+      <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$" />
+      <message key="name.invalidPattern"
+        value="Catch parameter name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="LocalVariableName">
+      <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$" />
+      <message key="name.invalidPattern"
+        value="Local variable name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="PatternVariableName">
+      <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$" />
+      <message key="name.invalidPattern"
+        value="Pattern variable name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="ClassTypeParameterName">
+      <property name="format" value="(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)" />
+      <message key="name.invalidPattern"
+        value="Class type name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="RecordComponentName">
+      <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$" />
+      <message key="name.invalidPattern"
+        value="Record component name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="RecordTypeParameterName">
+      <property name="format" value="(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)" />
+      <message key="name.invalidPattern"
+        value="Record type name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="MethodTypeParameterName">
+      <property name="format" value="(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)" />
+      <message key="name.invalidPattern"
+        value="Method type name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="InterfaceTypeParameterName">
+      <property name="format" value="(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)" />
+      <message key="name.invalidPattern"
+        value="Interface type name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="NoFinalizer" />
+    <module name="GenericWhitespace">
+      <message key="ws.followed"
+        value="GenericWhitespace ''{0}'' is followed by whitespace." />
+      <message key="ws.preceded"
+        value="GenericWhitespace ''{0}'' is preceded with whitespace." />
+      <message key="ws.illegalFollow"
+        value="GenericWhitespace ''{0}'' should followed by whitespace." />
+      <message key="ws.notPreceded"
+        value="GenericWhitespace ''{0}'' is not preceded with whitespace." />
+    </module>
+    <module name="Indentation">
+      <property name="basicOffset" value="2" />
+      <property name="braceAdjustment" value="2" />
+      <property name="caseIndent" value="2" />
+      <property name="throwsIndent" value="4" />
+      <property name="lineWrappingIndentation" value="4" />
+      <property name="arrayInitIndent" value="2" />
+    </module>
+    <module name="AbbreviationAsWordInName">
+      <property name="ignoreFinal" value="false" />
+      <property name="allowedAbbreviationLength" value="0" />
+      <property name="tokens"
+        value="CLASS_DEF, INTERFACE_DEF, ENUM_DEF, ANNOTATION_DEF, ANNOTATION_FIELD_DEF,
+                    PARAMETER_DEF, VARIABLE_DEF, METHOD_DEF, PATTERN_VARIABLE_DEF, RECORD_DEF,
+                    RECORD_COMPONENT_DEF" />
+    </module>
+    <module name="NoWhitespaceBeforeCaseDefaultColon" />
+    <module name="OverloadMethodsDeclarationOrder" />
+    <module name="VariableDeclarationUsageDistance" />
+    <module name="CustomImportOrder">
+      <property name="sortImportsInGroupAlphabetically" value="true" />
+      <property name="separateLineBetweenGroups" value="true" />
+      <property name="customImportOrderRules" value="STATIC###THIRD_PARTY_PACKAGE" />
+      <property name="tokens" value="IMPORT, STATIC_IMPORT, PACKAGE_DEF" />
+    </module>
+    <module name="MethodParamPad">
+      <property name="tokens"
+        value="CTOR_DEF, LITERAL_NEW, METHOD_CALL, METHOD_DEF,
+                    SUPER_CTOR_CALL, ENUM_CONSTANT_DEF, RECORD_DEF" />
+    </module>
+    <module name="NoWhitespaceBefore">
+      <property name="tokens"
+        value="COMMA, SEMI, POST_INC, POST_DEC, DOT,
+                    LABELED_STAT, METHOD_REF" />
+      <property name="allowLineBreaks" value="true" />
+    </module>
+    <module name="ParenPad">
+      <property name="tokens"
+        value="ANNOTATION, ANNOTATION_FIELD_DEF, CTOR_CALL, CTOR_DEF, DOT, ENUM_CONSTANT_DEF,
+                    EXPR, LITERAL_CATCH, LITERAL_DO, LITERAL_FOR, LITERAL_IF, LITERAL_NEW,
+                    LITERAL_SWITCH, LITERAL_SYNCHRONIZED, LITERAL_WHILE, METHOD_CALL,
+                    METHOD_DEF, QUESTION, RESOURCE_SPECIFICATION, SUPER_CTOR_CALL, LAMBDA,
+                    RECORD_DEF" />
+    </module>
+    <module name="OperatorWrap">
+      <property name="option" value="NL" />
+      <property name="tokens"
+        value="BAND, BOR, BSR, BXOR, DIV, EQUAL, GE, GT, LAND, LE, LITERAL_INSTANCEOF, LOR,
+                    LT, MINUS, MOD, NOT_EQUAL, PLUS, QUESTION, SL, SR, STAR, METHOD_REF,
+                    TYPE_EXTENSION_AND " />
+    </module>
+    <module name="AnnotationLocation">
+      <property name="id" value="AnnotationLocationMostCases" />
+      <property name="tokens"
+        value="CLASS_DEF, INTERFACE_DEF, ENUM_DEF, METHOD_DEF, CTOR_DEF,
+                      RECORD_DEF, COMPACT_CTOR_DEF" />
+    </module>
+    <module name="AnnotationLocation">
+      <property name="id" value="AnnotationLocationVariables" />
+      <property name="tokens" value="VARIABLE_DEF" />
+      <property name="allowSamelineMultipleAnnotations" value="true" />
+    </module>
+    <module name="NonEmptyAtclauseDescription" />
+    <module name="InvalidJavadocPosition" />
+    <module name="JavadocTagContinuationIndentation" />
+    <module name="SummaryJavadoc">
+      <property name="forbiddenSummaryFragments"
+        value="^@return the *|^This method returns |^A [{]@code [a-zA-Z0-9]+[}]( is a )" />
+    </module>
+    <module name="JavadocParagraph" />
+    <module name="RequireEmptyLineBeforeBlockTagGroup" />
+    <module name="AtclauseOrder">
+      <property name="tagOrder" value="@param, @return, @throws, @deprecated" />
+      <property name="target"
+        value="CLASS_DEF, INTERFACE_DEF, ENUM_DEF, METHOD_DEF, CTOR_DEF, VARIABLE_DEF" />
+    </module>
+    <module name="JavadocMethod">
+      <property name="accessModifiers" value="public" />
+      <property name="allowMissingParamTags" value="true" />
+      <property name="allowMissingReturnTag" value="true" />
+      <property name="allowedAnnotations" value="Override, Test" />
+      <property name="tokens" value="METHOD_DEF, CTOR_DEF, ANNOTATION_FIELD_DEF, COMPACT_CTOR_DEF" />
+    </module>
+    <module name="MissingJavadocMethod">
+      <property name="scope" value="public" />
+      <property name="minLineCount" value="2" />
+      <property name="allowedAnnotations" value="Override, Test" />
+      <property name="tokens"
+        value="METHOD_DEF, CTOR_DEF, ANNOTATION_FIELD_DEF,
+                                   COMPACT_CTOR_DEF" />
+    </module>
+    <module name="MissingJavadocType">
+      <property name="scope" value="protected" />
+      <property name="tokens"
+        value="CLASS_DEF, INTERFACE_DEF, ENUM_DEF,
+                      RECORD_DEF, ANNOTATION_DEF" />
+      <property name="excludeScope" value="nothing" />
+    </module>
+    <module name="MethodName">
+      <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9_]*$" />
+      <message key="name.invalidPattern"
+        value="Method name ''{0}'' must match pattern ''{1}''." />
+    </module>
+    <module name="SingleLineJavadoc" />
+    <module name="EmptyCatchBlock">
+      <property name="exceptionVariableName" value="expected" />
+    </module>
+    <module name="CommentsIndentation">
+      <property name="tokens" value="SINGLE_LINE_COMMENT, BLOCK_COMMENT_BEGIN" />
+    </module>
+    <!-- https://checkstyle.org/config_filters.html#SuppressionXpathFilter -->
+    <module name="SuppressionXpathFilter">
+      <property name="file" value="${org.checkstyle.google.suppressionxpathfilter.config}"
+        default="checkstyle-xpath-suppressions.xml" />
+      <property name="optional" value="true" />
+    </module>
+  </module>
+</module>

